libs/langgraph/src/managed/shared_value.ts

```diff
@@ -0,0 +1,122 @@
+import { RunnableConfig } from "@langchain/core/runnables";
+import { BaseStore, type Values } from "../store/base.js";
+import {
+  ConfiguredManagedValue,
+  ManagedValue,
+  ManagedValueParams,
+  WritableManagedValue,
+} from "./base.js";
+import { CONFIG_KEY_STORE } from "../constants.js";
+import { InvalidUpdateError } from "../errors.js";
+
+type Value = Record<string, Values>;
+type Update = Record<string, Values | null>;
+
+export interface SharedValueParams extends ManagedValueParams {
+  scope: string;
+  key: string;
+}
+
+export class SharedValue extends WritableManagedValue<Value, Update> {
+  scope: string;
+
+  store: BaseStore | null;
+
+  ns: string | null;
+
+  value: Value = {};
+
+  constructor(config: RunnableConfig, params: SharedValueParams) {
+    super(config, params);
+    this.scope = params.scope;
+    this.store = config.configurable?.[CONFIG_KEY_STORE] || null;
+
+    if (!this.store) {
+      this.ns = null;
+    } else if (config.configurable?.[this.scope]) {
+      const scopeValue = config.configurable[this.scope];
+      let scopedValueString = "";
+      try {
+        scopedValueString =
+          typeof scopeValue === "string"
+            ? scopeValue
+            : JSON.stringify(scopeValue);
+      } catch (_) {
+        // no-op
+      }
+      this.ns = `scoped:${this.scope}:${params.key}:${scopedValueString}`;
+    } else {
+      throw new Error(
+        `Scope ${this.scope} for shared state key not in config.configurable`
+      );
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static async initialize<Value = any>(
+    config: RunnableConfig,
+    args: SharedValueParams
+  ): Promise<ManagedValue<Value>> {
+    const instance = new this(config, args);
+    await instance.tick();
+    return instance as unknown as ManagedValue<Value>;
+  }
+
+  static on(scope: string): ConfiguredManagedValue<Value> {
+    return {
+      cls: SharedValue,
+      params: {
+        scope,
+        key: {},
```

Comment:
in py this is typed as a string, but defaults to an object. figure out which it should be.